### ISSUE TITLE
Correctly set libtool version info for all libraries

### ISSUE THREAD

**Issue by PCMan** [lxde/lxqt#488]
Currently, only libfm-qt follows the convention of libtool to set the version-info (so name).
Other libraries did not set it according to the convention of linux system.
Since cmake did not use libtool, I'd like to create some cmake macros to help handling the ABI version info.

**Comment by heliocastro** [comment]
CMake already do this, no need to reinvent the wheel:
Here's how we do on KDE

add_library( mylib SHARED )

set ( MYLIB_MAJOR_VERSION 0 )
set ( MYLIB_MINOR_VERSION 9 )
set ( MYLIB_PATCH_VERSION 0 )

set( MYLIB_VERSION
${MYLIB_MAJOR_VERSION}.${MYLIB_MINOR_VERSION}.${MYLIB_PATCH_VERSION} )

set_property( TARGET mylib PROPERTY VERSION  ${MYLIB_VERSION} )
set_property( TARGET mylib PROPERTY SOVERSION ${MYLIB_MAJOR_VERSION} )

And install so the proper symlink

**Comment by PCMan** [comment]
No, cmake did it differently from the convention of libtool. I studied this issue previously.
In libtool versioning scheme, current, revision, and age were used and it's the ABI version, which is not related to the actual version of the software package.
These values are used in different ways according to the OS you're using.
Libtools adjust the values behind your back while cmake only set whatever value it receives.
Apart from this, our lxqt libraries all set the major, minor, and patch versions to the software version.
However, according to libtool, the version-info/soname has different numbers and meanings from the software package version so they should not be set to the same values.

**Comment by palinek** [comment]
@PCMan what is needed to do here? IMO this is straight forward and everything that is needed is to un-tie the release version and the so/abi version. So in case of liblxqt e.g.:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4e159d5..6a029da 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -6,6 +6,9 @@ set(LXQT_MAJOR_VERSION 0)
 set(LXQT_MINOR_VERSION 10)
 set(LXQT_PATCH_VERSION 0)
 
+set(LIBLXQT_LIB_VERSION 1.0.0)
+set(LIBLXQT_LIB_SOVERSION 1)
+
 include(CMakePackageConfigHelpers)
 include(GNUInstallDirs) # Standard directories for installation
 
@@ -276,8 +279,8 @@ target_link_libraries(${LXQT_LIBRARY_NAME}
 )
 
 set_target_properties(${LXQT_LIBRARY_NAME} PROPERTIES
-    VERSION   ${LXQT_VERSION}
-    SOVERSION ${LXQT_MAJOR_VERSION}
+    VERSION   ${LIBLXQT_LIB_VERSION}
+    SOVERSION ${LIBLXQT_LIB_SOVERSION}
 )
 
 target_compile_definitions(${LXQT_LIBRARY_NAME}
```

No need for any cmake macro/function ... or am I missing something?

**Comment by luis-pereira** [comment]
@palinek CMake .so version scheme is different from libtool version scheme.

**Comment by palinek** [comment]
> CMake .so version scheme is different from libtool version scheme.

Please, be more specific or put some doc/link to read about it.

> Libtools adjust the values behind your back while cmake only set whatever value it receives.

@PCMan How is that?  I had a look in [libtool's doc](https://www.gnu.org/software/libtool/manual/libtool.html) and didn't found any "auto update". Actually based on [this paragraph](https://www.gnu.org/software/libtool/manual/libtool.html#Updating-version-info) I think that the managing the version info is full under control of the developer.

**Comment by luis-pereira** [comment]
@palinek Some pointers:
http://www.ibm.com/developerworks/linux/library/l-shlibs/index.html
https://cmake.org/Bug/bug_relationship_graph.php?bug_id=4383&graph=dependency

**Comment by paulolieuthier** [comment]
Can this be done for 0.11 (in the next few days)?

**Comment by paulolieuthier** [comment]
@luis-pereira what's your opinion?

**Comment by luis-pereira** [comment]
@paulolieuthier With commit lxde/libfm-qt@6f4782e12b9a1789a6d49e51e1f98e04ede2e896 we released 3.0.0. Source package [here](https://downloads.lxqt.org/libfm-qt/).
Versions are supposed to increase.

**Comment by paulolieuthier** [comment]
> @paulolieuthier With commit lxde/libfm-qt@6f4782e we released 3.0.0. Source package here.
Versions are supposed to increase.

I'm sure they are. I asked your opinion because I would like to know what you think about the use of libtool in LXQt, and whether you think we need to fix it before the next release, 0.11.

**Comment by luis-pereira** [comment]
@paulolieuthier It was kind of a joke, no pun intended, of course. I wasn't sure if you knew that we were in 3.0.0.
I'm not against using the libtool scheme.... if someone implements and maintains it.
From a consistency point of view, IMO we should use only one versioning scheme.

I don't see this as release blocker. Considering the changes since last release, 3.0.1 would be just fine.
Do you know why do we have an 0.11 tag already ?

**Comment by paulolieuthier** [comment]
> @paulolieuthier It was kind of a joke, no pun intended, of course. I wasn't sure if you knew that we were in 3.0.0.

Ah, OK. 😬 

> I'm not against using the libtool scheme.... if someone implements and maintains it.
From a consistency point of view, IMO we should use only one versioning scheme. I don't see this as release blocker.

I second all of that.

>  Do you know why do we have an 0.11 tag already ?

Don't know.

@luis-pereira It seems you're talking about libfm-qt only, and I'm talking about all LXQt packages. Anyway, I think we should move this issue to the next milestone and focus on the release blockers.

**Comment by agaida** [comment]
erm - maybe the tag is there, because libfm-qt and pcmanfm-qt was indeed released as 0.11 - and that was a long time ago (pcmanfm-qt: 2016-03-31 - Merge new version 0.11.0 from experimental to unstabledebian/0.11.0-2, libfm-qt: 2016-03-31- Merged changes into unstable debian/0.11.0-2).
So with the upcoming release we should release both packages at least at 0.11.1 - or 0.12. Only a reminder, pcmanfm-qt and libfm-qt are outside of the LXQt release cycle.

**Comment by luis-pereira** [comment]
@agaida Thanks for the explanation. A clarification: Versioning scheme doesn't mean that the actual version numbers are the same.

**Comment by agaida** [comment]
@luis-pereira - i know that, the only thing about is - are there any symbol changes that would make the new released lib not compatible with the current? if so i would suggest a bump to 4

**Comment by luis-pereira** [comment]
@agaida No new symbols since 0.11.0

**Comment by agaida** [comment]
BTW: The soname thing also apply to qtermwidget - we should make a bump and implement a versioning scheme for libqtermwidget too before release. Will file a bug.

**Comment by PCMan** [comment]
Sorry for being late on this thread. Let me try to clarify this.
@palinek: When I say libtool does something behind your back, here is an example.
The piece of code is taken from /usr/bin/libtool
`
        case $vinfo_number in
        yes)
          number_major=$1
          number_minor=$2
          number_revision=$3
          #
          # There are really only two kinds -- those that
          # use the current revision as the major version
          # and those that subtract age and use age as
          # a minor version.  But, then there is irix
          # that has an extra 1 added just for fun
          #
          case $version_type in
          # correct linux to gnu/linux during the next big refactor
          darwin|freebsd-elf|linux|osf|windows|none)
            func_arith $number_major + $number_minor
            current=$func_arith_result
            age=$number_minor
            revision=$number_revision
            ;;
          freebsd-aout|qnx|sunos)
            current=$number_major
            revision=$number_minor
            age=0
            ;;
          irix|nonstopux)
            func_arith $number_major + $number_minor
            current=$func_arith_result
            age=$number_minor
            revision=$number_minor
            lt_irix_increment=no
            ;;
          esac
          ;;
        no)
          current=$1
          revision=$2
          age=$3
          ;;
        esac
`
Clearly one can see that libtool parses the version-info string differently in different platforms. So the actual soname might not be the same as your version-info string.

I personally find the libtool version scheme a little bit annoying, but it seems to be the de facto standard in Linux and is used widely. Some distributions, for instance Debian/Ubuntu, name the packages containing libraries based on their sonames. If the sonames are not set correctly, there might be problems.
The idea of using different versioning for API and ABI can be cumbersome at first glance, but this actually makes sense. The correct and reliable dynamic linking rely on ABI version, not API versions.
API versioning is at the source code level. For a non-free commercial product which cannot be re-compiled against the new library whenever the API changes, correct ABI info is crucial to determine if the new library is compatible with the old executable binary, and vice versa.
Without ABI versions, this cannot be supported.
For our own dynamic libraries the lack of ABI versions is not problematic since we can re-compile the whole LXQt source tree. But if we're going to let third party developers use the libraries, then it's good practice to add correct ABI versions.
